tao_compiler/mlir/disc/transforms/disc_dot_merge.cc

```diff
@@ -0,0 +1,557 @@
+    }
+  };
+
+  struct DotShapeHash {
+    std::size_t operator()(const DotShape& dotShape) const {
+      std::size_t hash = dotShape.m_dim.hash();
+      hash = llvm::hash_combine(hash, dotShape.n_dim.hash());
+      hash = llvm::hash_combine(hash, dotShape.contracting_dim.hash());
+      for (const auto& d : dotShape.batching_dims) {
+        hash = llvm::hash_combine(hash, d.hash());
+      }
+      const auto& dimension_numbers = dotShape.dimension_numbers;
+      const auto& lhs_batch_dims = dimension_numbers.getLhsBatchingDimensions();
+      const auto& rhs_batch_dims = dimension_numbers.getRhsBatchingDimensions();
+      const auto& lhs_contracting_dims =
+          dimension_numbers.getLhsContractingDimensions();
+      const auto& rhs_contracting_dims =
+          dimension_numbers.getRhsContractingDimensions();
+      hash = llvm::hash_combine(hash,
+                                llvm::hash_combine_range(lhs_batch_dims.begin(),
+                                                         lhs_batch_dims.end()));
+      hash = llvm::hash_combine(hash,
+                                llvm::hash_combine_range(rhs_batch_dims.begin(),
+                                                         rhs_batch_dims.end()));
+      hash = llvm::hash_combine(
+          hash, llvm::hash_combine_range(lhs_contracting_dims.begin(),
+                                         lhs_contracting_dims.end()));
+      hash = llvm::hash_combine(
+          hash, llvm::hash_combine_range(rhs_contracting_dims.begin(),
+                                         rhs_contracting_dims.end()));
+      hash = llvm::hash_combine(hash, mlir::hash_value(dotShape.element_type));
+      return hash;
+    }
+  };
+
+  struct BatchCluster {
+    BatchCluster(Operation* op, int op_id) : leader_op_id(op_id) {
+      ops.push_back(op);
+    }
+
+    // Merges `other` into this cluster, and clears `other`.
+    void merge(BatchCluster& other) {
+      ops.insert(ops.end(), other.ops.begin(), other.ops.end());
+      other.ops.clear();
+    }
+
+    // ID of the representative node of this cluster.
+    int leader_op_id;
+
+    // Dot ops to be batched.
+    SmallVector<Operation*> ops;
+  };
+
+  bool buildShapedDotMap();
+  bool batchingDots();
+  Value expandDim0(OpBuilder& builder, Location& loc, Value value);
+  bool applyBatching(BatchCluster& cluster);
+
+ private:
+  FuncOp func_;
+  // A map to help to cluster dot that have same shape and dim-numbers together.
+  std::unordered_map<DotShape, SmallVector<mhlo::DotGeneralOp>, DotShapeHash>
+      equal_shape_dot_map_;
+};
+
+void DotBatchingConverter::run() {
+  if (!buildShapedDotMap()) {
+    return;
+  }
+  batchingDots();
+}
+
+bool DotBatchingConverter::buildShapedDotMap() {
+  ShapeAnalysis analysis(func_);
+  if (failed(analysis.run())) {
+    LLVM_DEBUG(llvm::dbgs() << "ShapeAnalysis failes for dot merge.\n");
+    return false;
+  }
+  func_.walk([&](mhlo::DotGeneralOp op) {
+    DotShape dot_shape;
```

Comment:
I don't quite understand the necessity of DotShape struct here.
By telling if two DotShape is equal, it's actually the same logic as requiring:
```
(*analysis.getShape(lhs()) == *analysis.getShape(other.lhs()) &&
*analysis.getShape(rhs()) == *analysis.getShape(other.rhs()) &&
dot_dimension_numbers() == other.dot_dimension_numbers())
```
is is neccessary to analyze b/m/n/k in separate here?

Reply:
We do not only want to know the equality, but also need a key to cluster the same-shaped dots together in a map. DotShape is the key. A DotOp with a customized comparator is still not good because their hash is different. The key requires all the dims can be expressed, which is the separated b/m/n/k here. Another reason that I use the dims separated is to enable the chance to batch some dots with extra reshapes in the future. Meanwhile, the planned ShapeAnalysisV2 combines DimValue and symbol-expr together, which makes the separate comparison between DimValues and the whole shape-expression the same.

Reply:
Will you do layout rewrite for two batchgemm if their b/m/n/ks are identical but has different layout?
If not, i still don't understand the necessity to analyze b/m/n/k in separate. To be discussed offline.